controller/engine_image_controller.go

```diff
@@ -378,9 +380,154 @@ func (ic *EngineImageController) syncEngineImage(key string) (err error) {
+			nodeName := v.Status.OwnerID
+			candidatePerNode[nodeName] = candidatePerNode[nodeName] + 1
+		}
+	}
+	// Only allow concurrentAutomaticEngineUpgradePerNodeLimit engines to be automatically upgraded at a time
+	for _, v := range upgradingCandidates {
+		node := v.Status.OwnerID
+		if candidatePerNode[node] < concurrentAutomaticEngineUpgradePerNodeLimit {
+			candidatePerNode[node] = candidatePerNode[node] + 1
+			finalCandidates = append(finalCandidates, v)
+		}
+	}
+	return finalCandidates
+}
+
+// getVolumesForEngineImageUpgrading returns a list of volumes that are qualified for engine image upgrading
+// A volume is qualified for engine image upgrading if it meets one of the following case:
+// Case 1:
+//   1. Volume is in detached state
+// Case 2:
+//   1. Volume is not in engine upgrading process
+//   2. Volume is in attached state and it is able to do live upgrade
+func (ic *EngineImageController) getVolumesForEngineImageUpgrading(volumes map[string]*longhorn.Volume, newEngineImageResource *longhorn.EngineImage) ([]*longhorn.Volume, error) {
+	candiates := []*longhorn.Volume{}
+
+	for _, v := range volumes {
+		// Already attempted to upgrade volume's engine image
+		if v.Spec.EngineImage == newEngineImageResource.Spec.Image {
+			continue
+		}
+		// Volume is in upgrading process. Need to wait for the upgrade to finish.
+		if v.Spec.EngineImage != v.Status.CurrentImage {
+			continue
+		}
+		if !ic.canDoOfflineEngineImageUpgrade(v, newEngineImageResource) && !ic.canDoLiveEngineImageUpgrade(v, newEngineImageResource) {
+			continue
+		}
+		candiates = append(candiates, v)
+	}
+
+	return candiates, nil
+}
+
+func (ic *EngineImageController) canDoOfflineEngineImageUpgrade(v *longhorn.Volume, newEngineImageResource *longhorn.EngineImage) bool {
+	return v.Status.State == types.VolumeStateDetached && newEngineImageResource.Status.State == types.EngineImageStateReady
+}
+
+// canDoLiveEngineImageUpgrade check if it is possible to do live engine upgrade for a volume
+// A volume can do live engine upgrade when:
+//   1. Volume is attached AND
+//   2. Volume's robustness is either healthy or degraded AND
+//   3. Volume's robustness is healthy if volume is DR volume
+//   4. Volume is not expanding AND
+//   5. The current volume's engine image is compatible with the new engine image
+func (ic *EngineImageController) canDoLiveEngineImageUpgrade(v *longhorn.Volume, newEngineImageResource *longhorn.EngineImage) bool {
+	if v.Status.State != types.VolumeStateAttached {
+		return false
+	}
+	if v.Status.Robustness != types.VolumeRobustnessHealthy && v.Status.Robustness != types.VolumeRobustnessDegraded {
+		return false
+	}
+	// TODO: Rebuild is not supported for old DR volumes and the handling of a degraded DR volume live upgrade will get stuck.
+	//  After all old DR volumes are gone, this check can be removed.
+	if v.Status.IsStandby && v.Status.Robustness != types.VolumeRobustnessHealthy {
```

Comment:
It should be OK not to upgrade a DR volume now. Since DR volumes must be activated before use and activating it would put it in the detached mode, it can be upgraded automatically in the detached mode.

Reply:
Thank you! Fixed the code and the LEP